libraries/chain/combined_database.cpp

```diff
@@ -0,0 +1,470 @@
+         kv_undo_stack(kv_database, rocksdb_undo_prefix) {}
+
+   void combined_database::set_backing_store(backing_store_type backing_store) {
+      if (backing_store == backing_store_type::ROCKSDB) {
+         if (db.get<kv_db_config_object>().backing_store == backing_store_type::ROCKSDB)
+            return;
+         auto& idx = db.get_index<kv_index, by_kv_key>();
+         auto  it  = idx.lower_bound(boost::make_tuple(kvdisk_id, name{}, std::string_view{}));
+         EOS_ASSERT(it == idx.end() || it->database_id != kvdisk_id, database_move_kv_disk_exception,
+                    "Chainbase already contains KV entries; use resync, replay, or snapshot to move these to "
+                    "rocksdb");
+         db.modify(db.get<kv_db_config_object>(), [](auto& cfg) { cfg.backing_store = backing_store_type::ROCKSDB; });
+      }
+      if (db.get<kv_db_config_object>().backing_store == backing_store_type::ROCKSDB)
+         ilog("using rocksdb for backing store");
+      else
+         ilog("using chainbase for backing store");
+   }
+
+   void combined_database::set_revision(uint64_t revision) {
+      try {
+         try {
+            db.set_revision(revision);
+#warning TODO: Chain_kv needs to fix kv_undo_stack's revision after restart
+            // kv_undo_stack.set_revision(revision, false);
+         }
+         FC_LOG_AND_RETHROW()
+      }
+      CATCH_AND_EXIT_DB_FAILURE()
+   }
+
+   void combined_database::undo() {
+      try {
+         try {
+            db.undo();
+            kv_undo_stack.undo(false);
+         }
+         FC_LOG_AND_RETHROW()
+      }
+      CATCH_AND_EXIT_DB_FAILURE()
+   }
+
+   void combined_database::commit(int64_t revision) {
+      try {
+         try {
+            db.commit(revision);
+            kv_undo_stack.commit(revision);
+         }
+         FC_LOG_AND_RETHROW()
+      }
+      CATCH_AND_EXIT_DB_FAILURE()
+   }
+
+   void combined_database::flush() {
+      try {
+         try {
+            kv_undo_stack.write_state();
+            kv_database.flush(true, true);
+         }
+         FC_LOG_AND_RETHROW()
+      }
+      CATCH_AND_EXIT_DB_FAILURE()
+   }
+
+   std::unique_ptr<kv_context> combined_database::create_kv_context(name receiver, kv_resource_manager resource_manager,
+                                                                    const kv_database_config& limits) {
+      switch (backing_store) {
+         case backing_store_type::ROCKSDB:
+            return create_kv_rocksdb_context(kv_database, kv_undo_stack, receiver, resource_manager, limits);
+         case backing_store_type::NATIVE: return create_kv_chainbase_context(db, receiver, resource_manager, limits);
+         default: EOS_ASSERT(false, action_validate_exception, "Unknown backing store.");
+      }
+   }
+
+   void combined_database::add_to_snapshot(
+         const eosio::chain::snapshot_writer_ptr& snapshot, const eosio::chain::block_state& head,
```

Comment:
this is a bit difficult to read.  I'm fine with punching these out to new lines and aligning the variable, but I would do it for all of them then.

Reply:
I'm using a .clang-format file, which I found in a sibling directory. The resulting formatting gives me the creeps. I think it's time to discuss the formatting rules that we use.